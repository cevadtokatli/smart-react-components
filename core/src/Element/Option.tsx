import styled from 'styled-components'
import applyStyledProps, { StyledProps, shouldForwardProp } from '../styled-props'

export default styled.option.withConfig({ shouldForwardProp })<StyledProps>(applyStyledProps)
