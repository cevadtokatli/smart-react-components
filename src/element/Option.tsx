import styled from "styled-components"
import {Props, attrs, css} from "./"

export default styled.option.attrs<Props>(attrs)<Props>(css)