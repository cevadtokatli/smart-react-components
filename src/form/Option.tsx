import React from "react"
import {SizeProps} from "../props/size-props"
import {JSXElementProps, JSXChild, Value, ValueMix, SetValue} from "../props"
import DV from "../default-value"
import OptionElement from "./OptionElement"

export interface Props extends SizeProps {
    elementProps?: JSXElementProps
    children: JSXChild
    value: Value
    searchValue?: string
    disabled?: boolean
    box?: boolean
    type?: string
    fill?: boolean
    shape?: string
    hover?: boolean
    space?: boolean
    active?: ValueMix
    setActive?: SetValue<ValueMix>
}

const Option: React.FC<Props> = ({size="default",sizeSm,sizeMd,sizeLg,sizeXl,elementProps=DV.JSX_ELEMENT_PROPS,children,value,disabled,box,type="primary",fill,shape="default",hover,active,setActive}) => {
    /**
     * Returns active$ status of the element. 
     */
    const getActive$ = (): boolean => {
        if(value == null)
            return false
        return Array.isArray(active) ? active.indexOf(value) > -1 : value == active
    }

    const [active$, setActive$] = React.useState<boolean>(() => getActive$())
    
    React.useEffect(() => setActive$(getActive$()), [value, active])

    /**
     * Sets value if element is not disabled.
     */
    const set = (): void => {
        if(disabled || (value == null && Array.isArray(active)))
            return

        if(Array.isArray(active)) {
            if(active.indexOf(value) == -1)
                setActive([...active,value])
            else
                setActive(active.filter(i => i != value))
        } else if(value != active)
            setActive(value)
    }

    return <OptionElement {...elementProps} className={`${active$ ? "src-option-active" : ""}`} size={size} sizeSm={sizeSm} sizeMd={sizeMd} sizeLg={sizeLg} sizeXl={sizeXl} box={box} type={type} fill$={fill} shape={shape} hover={hover} disabled={disabled} value={value} onClick={set}>{children}</OptionElement>
}

export default Option