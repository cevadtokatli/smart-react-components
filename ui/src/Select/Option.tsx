import { ContentElement, PaletteProp, SetState } from '@smart-react-components/core/types'
import React from 'react'
import DropdownItemElement from '../components/Dropdown/DropdownItemElement'
import { FormValue } from '../types'
import WaveEffect from '../WaveEffect'

export interface Props {
  children: ContentElement
  isDisabled?: boolean
  value: FormValue
}

export interface PrivateProps {
  active: FormValue | FormValue[]
  hasHover: boolean
  hasWaveEffect: boolean
  isDisabled: boolean
  isEmbedded: boolean
  isOutline: boolean
  isSoft: boolean
  palette: PaletteProp
  setActive: SetState<FormValue | FormValue[]>
  waveEffectPalette: PaletteProp
}

const Option: React.FC<Props> = ({ active, children, hasHover, hasWaveEffect, isDisabled, isOutline, isSoft, palette, setActive, value, waveEffectPalette }: (Props & PrivateProps)) => {
  const handleOnClick = () => {
    if (isDisabled) {
      return
    }

    if (!Array.isArray(active)) {
      setActive(value)
    } else {
      if (!active.includes(value)) {
        setActive([...active, value])
      } else {
        setActive(active.filter(i => i !== value))
      }
    }
  }

  const isActive = () => !Array.isArray(active) ? value === active : active.includes(value)

  const content = (
    <DropdownItemElement
      hasHover={hasHover}
      isActive={isActive()}
      isDisabled={isDisabled}
      isOutline={isOutline}
      isSoft={isSoft}
      palette={palette}
      onClick={handleOnClick}
    >
      {children}
    </DropdownItemElement>
  )

  return hasWaveEffect ? <WaveEffect palette={waveEffectPalette}>{content}</WaveEffect> : content
}

export default Option
