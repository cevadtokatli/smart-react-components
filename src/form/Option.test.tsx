import React from "react"
import renderTheme from "../test/renderTheme"
import mountTheme from "../test/mountTheme"
import Option from "./Option"

describe("<Option />", () => {
    it("should match snapshot", () => {
        const wrapper = renderTheme(<Option value="option-1">Option-1</Option>)
        expect(wrapper).toMatchSnapshot()
    })
    
    it("should include active class", () => {
        const wrapper = mountTheme(<Option value="option-1" active="option-1">Option-1</Option>)
        expect(wrapper.find(".src-option-active").exists()).toBe(true)
    })

    it("should not include active class", () => {
        const wrapper = mountTheme(<Option value="option-1" active="option-2">Option-1</Option>)
        expect(wrapper.find(".src-option-active").exists()).toBe(false)
    })

    it("should call setActive", () => {
        const setActive = jest.fn()
        const wrapper = mountTheme(<Option value="option-1" setActive={setActive}>Option-1</Option>)
        wrapper.find("div").simulate("click")
        expect(setActive).toHaveBeenCalled()
    })

    it("should not call setActive", () => {
        const setActive = jest.fn()
        const wrapper = mountTheme(<Option value="option-1" setActive={setActive} disabled={true}>Option-1</Option>)
        wrapper.find("div").simulate("click")
        expect(setActive).not.toHaveBeenCalled()
    })
})