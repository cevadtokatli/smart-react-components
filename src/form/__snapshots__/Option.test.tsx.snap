// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Option /> should match snapshot 1`] = `
<div
  class="sc-AxjAm sc-AxirZ eGnNeZ"
  shape="default"
  type="primary"
  value="option-1"
>
  Option-1
</div>
`;
